qiskit/circuit/__init__.py

```diff
@@ -54,81 +78,825 @@
+:class:`QuantumCircuit` interface is designed for storing and building circuits, while the
+:ref:`transpiler and its passes <qiskit-transpiler>`, and its intermediate :class:`.DAGCircuit`
+representation, are where you should look for an interface to mutate circuits.
+
+The :class:`QuantumCircuit` methods that add instructions to circuits (such as
+:meth:`~QuantumCircuit.append`, and all the helper standard-gate methods) return an
+:class:`InstructionSet`, which is a handle to several :class:`CircuitInstruction`\ s simultaneously.
+
+.. autoclass:: InstructionSet
+    :members:
+
+This :class:`InstructionSet` is now little used in Qiskit.  It provides a very minimal set of
+methods to perform post-append mutations on instructions (which *will* be propagated to the
+circuit), but these are now discouraged and you should use the alternatives noted in those methods.
+
+
+Operations, instructions and gates
+----------------------------------
+
+Within a :class:`CircuitInstruction`, the minimal interface that any operation must fulfil is
+:class:`Operation`.  This is a *very* high level view, and only usable for abstract circuits.  The
+main purpose of treating operations as :class:`Operation` is to allow arbitrary mathematical
+objects (such as :class:`.quantum_info.Operator`) to be added to abstract circuits directly.
+
+.. autoclass:: Operation
+    :members:
+
+Most operations, including all operations on physical circuits, are instances of the more concretely
+defined :class:`Instruction`.  This represents any instruction that some QPU might be able to carry
+out natively, such as :class:`Measure`.  :class:`Instruction` need not be unitary (much as
+:class:`Measure` isn't); an instruction is specifically unitary if it is a :class:`Gate`.
+
+.. autoclass:: Instruction
+    :members:
+    :show-inheritance:
+
+:class:`Instruction`\ s can be near arbitrary, provided they only act on :class:`Qubit`\ s and
+:class:`Clbit`\ s, and are parametrized by their :attr:`~Instruction.params`; they should not
+attempt to "close over" outer circuit registers, or use hidden parameters inside themselves.
+:class:`Instruction`\ s can be related to other circuits to provide a decompositions by using
+their :attr:`Instruction.definition` attribute, which provides a local, one-off decomposition.  This
+can be in whatever basis set of operations is most convenient to you, as long as the definitions of
+all contained gates have some topological order; that is, you cannot use a gate in a definition it
+its own definition depends on the parent.  If the :class:`Instruction` should be considered entirely
+opaque to optimizers, its :attr:`~Instruction.definition` can be ``None``.  See
+:ref:`circuit-custom-gates` for more detail.
+
+The :attr:`~Instruction.params` of an instruction can technically be arbitrary, but in general you
+should attempt to stick to parametrizations in terms of angles, wherever possible.  Qiskit itself
```

Comment:
Can we generalize 'angles' to reals?

Reply:
We probably can. Angles were what these were originally meant to be (with an intent that control hardware might want to treat them with fixed-point integer semantics), but in practice, real numbers are probably fine.

Reply:
Sounds good! :)

Reply:
Done in 2c1cc17232a.